backend/azure/dtmb/internal/utils/datastructures.go

```diff
@@ -0,0 +1,144 @@
+package utils
+
+import (
+	"container/list"
+	"sync"
+
+	dtmbprotos "github.com/microsoft/durabletask-go/backend/azure/dtmb/internal/backend/v1"
+)
+
+const (
+	DefaultOrchestrationHistoryCacheSize int = 100
+)
+
+type SyncQueue[T dtmbprotos.ExecuteOrchestrationMessage | dtmbprotos.ExecuteActivityMessage | dtmbprotos.Event] struct {
+	lock  *sync.Mutex
+	items []*T
+}
+
+func NewSyncQueue[T dtmbprotos.ExecuteOrchestrationMessage | dtmbprotos.ExecuteActivityMessage | dtmbprotos.Event]() SyncQueue[T] {
+	return SyncQueue[T]{
+		lock:  &sync.Mutex{},
+		items: []*T{},
+	}
+}
+
+func (q *SyncQueue[T]) Enqueue(item *T) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.items = append(q.items, item)
+}
+
+func (q *SyncQueue[T]) EnqueueBatch(items []*T) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.items = append(q.items, items...)
+}
+
+func (q *SyncQueue[T]) Dequeue() *T {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	if len(q.items) == 0 {
+		return nil
+	}
+
+	item := q.items[0]
+	q.items[0] = nil // avoid a leak
+	q.items = q.items[1:]
+	return item
+}
+
+func (q *SyncQueue[T]) ReadAllWithoutDequeue() []*T {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	return q.items
+}
+
+func (q *SyncQueue[T]) ReadLastWithoutDequeue() *T {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	if len(q.items) == 0 {
+		return nil
+	}
+	return q.items[len(q.items)-1]
+}
+
+type OrchestrationHistoryCache struct {
+	lock *sync.Mutex
+
+	capacity int
+	cache    map[string]*list.Element
+	list     *list.List
```

Comment:
What's the reason for using this data type?

Reply:
LRU implementation! We can easily move a list element (which is a particular orchestration history stack) to the front of the list (meaning used most recently). If we exceed our capacity (list length) we can remove the last element in the list.

There is some synergy between the list Elements and list length.

Reply:
Oh great!